### ISSUE TITLE
[Bug]: No Resolution switching with refresh rate

### ISSUE THREAD

**Issue by KyleSanderson** [nova-video-player/aos-AVP#625]
### Problem description

Resolution is still 4k... doesn't look like the refresh rate changed either.

### Steps to reproduce the issue

Turn refresh rate switching on.

### Expected behavior

Display on my television should be 1080/24hz.

### Your phone/tablet/androidTV model

Firestick Max

### Operating system version

AMZN

### Application version and app store

6.0.62

### Additional system information

Sony

**Comment by courville** [comment]
This is correct, the current algorithms pick the "best" refresh rate for a given resolution. This is a limitation especially with 4k display mode with reduced refresh rate set. I propose to take the highest resolution above or equal 1080p matching the best refresh rate when video resolution is not 4k.
I would be interested to know the supported rates of your TV though to verify that 24fps is not supported in 4k.

**Comment by courville** [comment]
OK after some experiments here are some results.
I first need to correct an error in the refresh rate selection due to some rounding errors happening only on firestick.
Moreover, in order to get adaptive refresh rate on firestick you need to enable it in fireOS settings -> display & sounds -> Display, put video resolution to auto (up to 4k ultra hd); -> match original frame rate ON; -> dynamic range settings disable HDR or set it to adaptive. If you do not do this you will only have a restrictive set of refresh rates.
Note that choosing adaptive for dynamic range setting, your UI will be 1080p, and disabled will enable a 4k UI.
With this setting I observe that all the refresh rates are available both for 1080p and 4k screen resolution.
I thus conclude that there is no need for resolution adaptation in the refresh rate selection.
And I need to fix a bug to get it fully working on firestick.

TODO: put this observation in FAQ.

**Comment by courville** [comment]
@KyleSanderson to simplify the message, I see no need to change resolution to get a better refresh rate.

**Comment by KyleSanderson** [comment]
> @KyleSanderson to simplify the message, I see no need to change resolution to get a better refresh rate.

This is the mistake that Kodi made that they have since corrected in 2018. TV upscalers are much more correct than using what's on the player. This is because the TV itself knows the native resolution of the panel, and has enhanced features to improve image quality if it knows the original source is 720p or 1080p on a 4K panel.

@courville refresh rate is there for tearing, resolution changing is there for picture quality. Both are part of seamless quality video playback. Kodi on Android (firestick included) has had this resolution switching as well for quite some time.

> Moreover, in order to get adaptive refresh rate on firestick you need to enable it in fireOS settings -> display & sounds -> Display, put video resolution to auto (up to 4k ultra hd); -> match original frame rate ON; -> dynamic range settings disable HDR or set it to adaptive. If you do not do this you will only have a restrictive set of refresh rates.

I do have those settings on my firestick - Kodi behaves correctly for most media with switching.

The feature merge is here from Kodi: https://github.com/xbmc/xbmc/pull/13274

**Comment by courville** [comment]
I understand your point of having TV upscalers of better quality than the Android hwcodecs ones.
There will be the question of is it any better than the AI upscaling on nvidia shield (hardware fragmentation).
I am against introducing zillions of settings in the app to cover all the corner cases.
I will think about it.

**Comment by KyleSanderson** [comment]
I'm on the latest amzn build that's been pushed but I'm still not seeing the expected /24hz on the television (it could be it doesn't display properly on 23.976). Is there a way to view what mode / settings aos is expecting to be set?

This is what my panel is capable of re: Firestick.

`04-24 20:16:21.149   545   571 I DisplayManagerService: Display device changed: DisplayDeviceInfo{"Built-in Screen": uniqueId="local:0", 3840 x 2160, modeId 11, defaultModeId 1, supportedModes [{id=1, width=1920, height=1080, fps=60.000004}, {id=2, width=1920, height=1080, fps=59.94}, {id=3, width=3840, height=2160, fps=59.94}, {id=4, width=1280, height=720, fps=59.94}, {id=5, width=3840, height=2160, fps=50.0}, {id=6, width=3840, height=2160, fps=60.000004}, {id=7, width=3840, height=2160, fps=30.0}, {id=8, width=3840, height=2160, fps=29.97}, {id=9, width=3840, height=2160, fps=25.0}, {id=10, width=3840, height=2160, fps=24.0}, {id=11, width=3840, height=2160, fps=23.976}, {id=12, width=1920, height=1080, fps=50.0}, {id=13, width=1920, height=1080, fps=29.97}, {id=14, width=1920, height=1080, fps=23.976}, {id=15, width=1920, height=1080, fps=24.0}, {id=16, width=1920, height=1080, fps=30.0}, {id=17, width=1280, height=720, fps=50.0}, {id=18, width=1280, height=720, fps=60.000004}, {id=19, width=720, height=576, fps=50.0}, {id=20, width=720, height=480, fps=60.000004}], colorMode 0, supportedColorModes [0], HdrCapabilities android.view.Display$HdrCapabilities@bf1c8d68, density 320, 314.632 x 322.729 dpi, appVsyncOff 1000000, presDeadline 41708375, touch INTERNAL, rotation 0, type BUILT_IN, state ON, FLAG_DEFAULT_DISPLAY, FLAG_SECURE, FLAG_SUPPORTS_PROTECTED_BUFFERS}`

![image](https://user-images.githubusercontent.com/1184902/165015742-dfd0fd9c-67df-4b1c-9236-20fc7e6f100a.png)
 attached image is what is actually sent to the firestick via extended edid.

**Comment by courville** [comment]
Just checked a 24fps video that is displayed correctly on my TV.
You should see it in nova logs as well (I enabled the Player logs for this release).
You can use `adb shell am start com.amazon.ssm/com.amazon.ssm.ControlPanel` as well to check.
Please report nova player logs to see what is going on.
